### PR TITLE
Fix: Elasticsearch domains should be in a VPC (#4688)

### LINKED ISSUE #4688: Elasticsearch domains should be in a VPC
```
{
    "GeneratorIds": [
        "aws-foundational-security-best-practices/v/1.0.0/ES.2"
    ]
}
```

https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-standards-fsbp-controls.html#fsbp-es-2

---

- [ ] Security design review completed; the Resolution of this issue does **not** …
  - [ ] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [ ] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [ ] … affect the permissions of external users like access to
    - TDR snapshots
  - [ ] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [ ] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [ ] … affect monitoring of the system
  - [ ] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [ ] … add an interface that exposes sensitive or confidential data at the security boundary
  - [ ] … affect the encryption of data at rest
  - [ ] … require persistence of sensitive or confidential data that might require encryption at rest
  - [ ] … require unencrypted transmission of data within the security boundary
  - [ ] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [ ] Documentation on any unchecked boxes is provided in comments below

### PR DESCRIPTION
**Pull request by nadove-ucsc**
Connected issues: #4688

### Author

- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue        <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via Zenhub 
- [x] PR description links to connected issues
- [x] Added `partial` label to PR                                   <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR title is `Fix: ` followed by the issue title   

### Author (reindex)

- [x] Added `r` tag to commit title                                 <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                                   <sub>or this PR does not require reindexing</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain               <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR                        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR                             <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst          <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                                 <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                                   <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks           <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title                            <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues      <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                                <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                                 <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                                      <sub>or this PR does not touch requirements*.txt</sub>

### Author (rebasing, integration test)

- [x] `make integration_test` passes in personal deployment         <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased PR branch on `develop`, squashed old fixups

### Peer reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists.

### Peer reviewer (after approval)

- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists. Update the `N reviews` label. 

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations         <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear                      <sub>or all connected issues are labeled `no demo`</sub>
- [x] Rebased and squashed PR branch
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Followed notification steps in UPGRADING.rst for `sandbox`
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label    <sub>or PR is labeled `no sandbox`</sub>
- [x] Followed steps in UPGRADING.rst for `sandbox`
- [x] Build passes in `sandbox` deployment                          <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build log for anomalies in `sandbox` deployment      <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox`                     <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `sandbox`                                  <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                             <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Followed notification steps in UPGRADING.rst for `anvilbox`
- [x] Pushed PR branch to GitLab `anvildev`                         <sub>or PR is labeled `no sandbox`</sub>
- [x] Followed steps in UPGRADING.rst for `anvilbox`
- [x] Build passes in `anvilbox` deployment                         <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build log for anomalies in `anvilbox` deployment     <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `anvilbox`                    <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `anvilbox`                                 <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox`                            <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issues to Merged column
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [x] Shortened the PR chain                                        <sub>or this PR is not labeled `base`</sub>
- [x] Pushed merge commit to GitLab `dev`                           <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvildev`                      <sub>or PR is labeled `no sandbox`</sub>
- [x] Followed steps in UPGRADING.rst for `dev`
- [x] Followed steps in UPGRADING.rst for `anvildev`
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build log for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build log for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is* 
pushed determines this checklist item.

### Operator (reindex) 

- [x] Deleted unreferenced indices in `dev`                         <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Deleted unreferenced indices in `anvildev`                    <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `dev`                                      <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev`                                 <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev`            <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev`       <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment                       <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment                  <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Added upgrade-related checkbox items to promotion PR
- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

File: terraform/gitlab/gitlab.tf.json.template.py [hunk 1]
```diff
@@ -696,13 +696,17 @@ def qq(*words):
...
-            f'gitlab_{vpc.subnet_name(public)}_{zone}': {
+            f'{component}_{vpc.subnet_name(public)}_{zone}': {
                 'availability_zone': f'${{data.aws_availability_zones.available.names[{zone}]}}',
-                'cidr_block': f'${{cidrsubnet(aws_vpc.gitlab.cidr_block, 8, {vpc.subnet_number(zone, public)})}}',
+                'cidr_block': f'${{cidrsubnet(aws_vpc.gitlab.cidr_block, 8, {vpc.subnet_number(component, zone, public)})}}',
                 'map_public_ip_on_launch': public,
                 'vpc_id': '${aws_vpc.gitlab.id}'
             }
-            for public in (False, True)
+            for (component, public) in [
```
Comment (hannes-ucsc) [inline terraform/gitlab/gitlab.tf.json.template.py:705]: I don't think you need to add a third subnet. The ES cluster should go into the existing private subnet.

File: terraform/gitlab/gitlab.tf.json.template.py [hunk 2]
```diff
@@ -878,6 +882,16 @@ def qq(*words):
                                       from_port=3,  # Destination Unreachable
                                       to_port=4)  # Fragmentation required DF-flag set
                 ]
+            },
+            'elasticsearch': {
```
Comment (hannes-ucsc) [inline terraform/gitlab/gitlab.tf.json.template.py:886]: Since there is no reference to this in the `gitlab` component this can be moved to the main component. Generally speaking, you should minimize the number of resources added to the main component.

**Review by hannes-ucsc — COMMENTED** [review]
Unsolicited WIP review.

File: terraform/elasticsearch.tf.json.template.py [hunk 3]
```diff
@@ -65,19 +66,6 @@
...
-                                "Principal": {
-                                    "AWS": "*"
-                                },
-                                "Action": "es:*",
-                                "Resource": "arn:aws:es:${local.region}:${local.account_id}:domain/" + domain + "/*",
-                                "Condition": {
-                                    "IpAddress": {
-                                        "aws:SourceIp": []
-                                    }
-                                }
```
Comment (hannes-ucsc) [inline terraform/elasticsearch.tf.json.template.py:80]: This hunk should be a separate commit titled "Remove obsolete IAM policy section allowing access to ES by IP"

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Be sure to test the kibana_proxy.py script.

```diff
Index: src/azul/service/lambda_iam_policy.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/service/lambda_iam_policy.py b/src/azul/service/lambda_iam_policy.py
--- a/src/azul/service/lambda_iam_policy.py	(revision 7e00b0acb5f30c3eb6fd532cbda3bbcc3be3abea)
+++ b/src/azul/service/lambda_iam_policy.py	(date 1668124354245)
@@ -2,7 +2,7 @@
     config,
 )
 from azul.chalice import (
-    vpc_policy,
+    vpc_lambda_iam_policy,
 )
 from azul.deployment import (
     aws,
@@ -183,6 +183,6 @@
             ] if direct_access_role is not None else [
             ]
         ),
-        *vpc_policy()
+        *vpc_lambda_iam_policy()
     ]
 }
Index: terraform/modules.tf.json.template.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/terraform/modules.tf.json.template.py b/terraform/modules.tf.json.template.py
--- a/terraform/modules.tf.json.template.py	(revision 7e00b0acb5f30c3eb6fd532cbda3bbcc3be3abea)
+++ b/terraform/modules.tf.json.template.py	(date 1668124354233)
@@ -32,11 +32,12 @@
                 f"${{data.aws_subnet.gitlab_{vpc.subnet_name(public=False)}_{zone}.id}}"
                 for zone in range(vpc.num_zones)
             ]),
-            **({
-                   config.var_vpc_endpoint_id: f"${{aws_vpc_endpoint.{lambda_name}.id}}"
-               } if config.private_api else {
-
-            }),
+            **(
+                {
+                    config.var_vpc_endpoint_id: f"${{aws_vpc_endpoint.{lambda_name}.id}}"
+                } if config.private_api else {
+                }
+            ),
         } for lambda_name in config.lambda_names()
     }
 })
Index: src/azul/chalice.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/chalice.py b/src/azul/chalice.py
--- a/src/azul/chalice.py	(revision 7e00b0acb5f30c3eb6fd532cbda3bbcc3be3abea)
+++ b/src/azul/chalice.py	(date 1668124354251)
@@ -441,7 +441,7 @@
     }
 
 
-def vpc_policy(for_tf: bool = False):
+def vpc_lambda_iam_policy(for_tf: bool = False):
     """
     Returns the fragment of IAM policy JSON needed for placing a Lambda function
     into a VPC.
Index: src/azul/indexer/lambda_iam_policy.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/indexer/lambda_iam_policy.py b/src/azul/indexer/lambda_iam_policy.py
--- a/src/azul/indexer/lambda_iam_policy.py	(revision 7e00b0acb5f30c3eb6fd532cbda3bbcc3be3abea)
+++ b/src/azul/indexer/lambda_iam_policy.py	(date 1668124354236)
@@ -2,7 +2,7 @@
     config,
 )
 from azul.chalice import (
-    vpc_policy,
+    vpc_lambda_iam_policy,
 )
 from azul.deployment import (
     aws,
@@ -125,6 +125,6 @@
             ] if direct_access_role is not None else [
             ]
         ),
-        *vpc_policy()
+        *vpc_lambda_iam_policy()
     ]
 }
Index: terraform/gitlab/gitlab.tf.json.template.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/terraform/gitlab/gitlab.tf.json.template.py b/terraform/gitlab/gitlab.tf.json.template.py
--- a/terraform/gitlab/gitlab.tf.json.template.py	(revision 7e00b0acb5f30c3eb6fd532cbda3bbcc3be3abea)
+++ b/terraform/gitlab/gitlab.tf.json.template.py	(date 1668124354242)
@@ -21,7 +21,7 @@
     ServiceActionType,
 )
 from azul.chalice import (
-    vpc_policy,
+    vpc_lambda_iam_policy,
 )
 from azul.collections import (
     dict_merge,
@@ -465,7 +465,7 @@
                                    UUID='*',
                                    LayerVersion='*'),
 
-                    *vpc_policy(for_tf=True),
+                    *vpc_lambda_iam_policy(for_tf=True),
 
                     # CloudWatch does not describe any resource-level
                     # permissions
Index: terraform/api_gateway.tf.json.template.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/terraform/api_gateway.tf.json.template.py b/terraform/api_gateway.tf.json.template.py
--- a/terraform/api_gateway.tf.json.template.py	(revision 7e00b0acb5f30c3eb6fd532cbda3bbcc3be3abea)
+++ b/terraform/api_gateway.tf.json.template.py	(date 1668124354231)
@@ -476,7 +476,7 @@
                             lambda_.name: {
                                 "vpc_endpoint_id": "${aws_vpc_endpoint.%s.id}" % lambda_.name,
                             }
-                        },
+                        }
                     }
                     if config.private_api else {
                     }
Index: scripts/prepare_lambda_deployment.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/scripts/prepare_lambda_deployment.py b/scripts/prepare_lambda_deployment.py
--- a/scripts/prepare_lambda_deployment.py	(revision 7e00b0acb5f30c3eb6fd532cbda3bbcc3be3abea)
+++ b/scripts/prepare_lambda_deployment.py	(date 1668124354247)
@@ -36,10 +36,12 @@
         'cloudwatch_log_group_provisioner': {},
         config.var_vpc_subnet_ids: {},
         config.var_vpc_security_group_id: {},
-        **({
-               config.var_vpc_endpoint_id: {},
-           } if config.private_api else {
-        })
+        **(
+            {
+                config.var_vpc_endpoint_id: {},
+            } if config.private_api else {
+            }
+        )
     }
 
     input_json['output']['stage_name'] = {
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Oh, and there is an extra `i` in the commit title.

**Comment by hannes-ucsc** [comment]
Cancelling peer review due to urgency.

**Review by hannes-ucsc — APPROVED** [review]

**Review by dsotirho-ucsc — DISMISSED** [review]

File: scripts/kibana_proxy.py [hunk 4]
```diff
@@ -146,7 +146,7 @@ def es_endpoint(self):
         log.info('Getting domain endpoint')
         es = aws.es
         domain = es.describe_elasticsearch_domain(DomainName=self.options.domain)
-        return 'https://' + domain['DomainStatus']['Endpoint']
+        return 'https://' + domain['DomainStatus']['Endpoints']['vpc']
```
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply scripts/kibana_proxy.py:149]: But @danielsotirhos, this hasn't been deployed yet to `dev` so obviously the new code won't work against it.

File: src/azul/deployment.py [hunk 5]
```diff
@@ -192,7 +192,7 @@ def es_endpoint(self) -> Optional[Netloc]:
         if config.es_endpoint:
             return config.es_endpoint
         else:
-            return self._es_domain_status['Endpoint'], 443
+            return self._es_domain_status['Endpoints']['vpc'], 443
```
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply src/azul/deployment.py:195]: Same.

File: terraform/elasticsearch.tf.json.template.py [hunk 6]
```diff
@@ -134,6 +131,18 @@
                         ]
                     }
                 }
+            },
+            "aws_security_group": {
+                "elasticsearch": {
+                    "name": "azul-elasticsearch",
```
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply terraform/elasticsearch.tf.json.template.py:137]: This makes sense.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Please address 

https://github.com/DataBiosphere/azul/pull/4725#discussion_r1021948136

from @danielsotirhos's review. His other two change requests don't make sense, I think.

**Review by hannes-ucsc — DISMISSED** [review]

**Comment by noah-aviel-dove** [comment]
IT passed after following upgrading steps for personal deployment with non-shared ES domain.
